### ISSUE TITLE
MemoryError when trying to convert BSP format 11

### ISSUE THREAD

**Issue by LukerMaster** [frankvollebregt/GBSPConverter#7]
I have a BSP map file from a game "The Italian Job" published in 2001 and I tried to use your script to convert it into OBJ and IBSP.
In both cases I get the exact same error:
```
C:\Users\Luker\Downloads\GBSPConverter-src-import-fix\GBSPConverter-src-import-fix\src>python gbsp_to_ibsp.py track.bsp
Traceback (most recent call last):
  File "C:\Users\Luker\Downloads\GBSPConverter-src-import-fix\GBSPConverter-src-import-fix\src\gbsp_to_ibsp.py", line 35, in <module>
    gbsp = read_gbsp_file(path)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Luker\Downloads\GBSPConverter-src-import-fix\GBSPConverter-src-import-fix\src\gbsp_to_ibsp.py", line 20, in read_gbsp_file
    current_chunk = read_gbsp_chunk(file)
                    ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Luker\Downloads\GBSPConverter-src-import-fix\GBSPConverter-src-import-fix\src\gbsp_to_ibsp.py", line 12, in read_gbsp_chunk
    chunk.read_bytes(f)
  File "C:\Users\Luker\Downloads\GBSPConverter-src-import-fix\GBSPConverter-src-import-fix\src\gbsp.py", line 24, in read_bytes
    self.bytes = file.read(self.size * self.elements)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
MemoryError
```
BSP version (according to specification) is 11 and I could open it with in RWanalyze (however this program does not allow preview, editing, and bsp->dff conversion does not seem to work for this file).

I understand that this may not be the issue with the script itself since these particular BSP files are quite a problematic case as none of the methods for opening BSPs listed online currently work for them.

The specific file tested I used is [here](https://github.com/frankvollebregt/GBSPConverter/files/14899515/leveld.zip).
The game (alongside all the bsp files located in gfx/land directory) is also downloadable [here](https://www.myabandonware.com/game/the-italian-job-des).

**Comment by frankvollebregt** [comment]
Hey! Nice to see someone trying to use these scripts :) I'll have a look (probably later today or tomorrow) to look into this for a bit

**Comment by frankvollebregt** [comment]
Hello! I'm afraid this converter is not gonna be able to handle files from _The Italian Job_. It was made specifically to process BSP files from games written in the Genesis3D game engine (hence the name GBSP for Genesis BSP), which uses a derivative of the Quake Engine's BSP format.

I guess (but don't quote me on this) the BSP file used by _The Italian Job_ uses the same Binary Space Partitioning technique for efficient rendering that gives BSP its name, but it seems to be written for and working with a completely different engine.

Looking on [the game's Wikipedia page](https://en.wikipedia.org/wiki/The_Italian_Job_(2001_video_game)), it's mentioned that the game uses the [Renderware engine](https://en.wikipedia.org/wiki/RenderWare).

Interestingly, I found [this page on gtamods.com](https://gtamods.com/wiki/RenderWare_binary_stream_file) mentioning the fact that different file extensions have been used for this engine. Moreover, I found a site (can't find the URL right now) that mentions the `.bsp` extension has been replaced with `.rws` for later versions of the engine, but that the content has remained the same.

Thus, you might have some luck loading your BSP file into a tool that was written to read or manipulate `.rws` files used by the Renderware engine.

**Comment by frankvollebregt** [comment]
I will add that the Renderware engine had a lot more people using it and games being produced with it, so I would assume there is a tool to use out there somewhere ;)

**Comment by frankvollebregt** [comment]
Using RW Analyze (made mostly for GTA III, VC, SA), I can open and see some content of the BSP file you linked.

![afbeelding](https://github.com/frankvollebregt/GBSPConverter/assets/23523399/67f41f51-bc3d-4c33-aad6-b52e6a46b7f6)

http://steve-m.com/downloads/tools/rwanalyze/

I hope I've given you some ideas on what to look for 😄 if I figure anything else out to read these files I'll let you know here

**Comment by LukerMaster** [comment]
Thanks a lot for all the help. Note on rws files turned out to be really useful. I eventually found [this](https://www.vg-resource.com/thread-21139.html) website that lead me to [BSP2OBJ](http://www.mediafire.com/download/f48uwdos76gasx6/bsp2objv021(sonic_pc_support).exe) program which (somewhat) successfully extracted the map after some fiddling with the settings. [Magic.TXD](https://www.gtagarage.com/mods/show.php?id=27862) then allowed me to extract textures, and after semi-automatic merging in blender, I got the map.

I guess that is it for this issue then, unless you'd like to implement RenderWare map support in the future 😄. 

Edit: This tool does not seem to be working fully. I got some of the maps, however of the two most important ones, one of them is read as a single vertex and the other is missing most of its parts. Better than nothing though.
